dwave/optimization/src/nodes/set_routines.cpp

```diff
@@ -0,0 +1,254 @@
+    add_predecessor(element_ptr);
+    add_predecessor(test_elements_ptr);
+}
+
+double const* IsInNode::buff(const State& state) const {
+    return data_ptr<IsInNodeData>(state)->buff();
+}
+
+void IsInNode::commit(State& state) const { data_ptr<IsInNodeData>(state)->commit(); }
+
+std::span<const Update> IsInNode::diff(const State& state) const {
+    return data_ptr<IsInNodeData>(state)->diff();
+}
+
+void IsInNode::initialize_state(State& state) const {
+    const Array::View element = element_ptr_->view(state);
+    const Array::View test_elements = test_elements_ptr_->view(state);
+
+    emplace_data_ptr<IsInNodeData>(state, std::vector<double>{element.begin(), element.end()},
+                                   std::vector<double>{test_elements.begin(), test_elements.end()});
+}
+
+bool IsInNode::integral() const { return true; }  // All values are true/false
+
+double IsInNode::min() const { return 0.0; }  // All values are true/false
+
+double IsInNode::max() const { return 1.0; }  // All values are true/false
+
+inline void IsInNode::rm_index(IsInNodeData*& node_data, const ssize_t rm_index,
+                               const double key) const {
+    auto set_data_it = node_data->set_data.find(key);
+    // find() should be successful since we need to remove the index
+    assert(set_data_it != node_data->set_data.end());
+    // Vector containing indices of `element` with value `key`
+    std::vector<ssize_t>& indices_vec = set_data_it->second.element_indices;
+    // Find index of `indices_vec` equal to `rm_index`
+    auto index_ptr = std::find(indices_vec.begin(), indices_vec.end(), rm_index);
+    // find() should be successful since we need to remove the index
+    assert(index_ptr != indices_vec.end());
+    // swap and pop
+    *index_ptr = indices_vec.back();
+    indices_vec.pop_back();
+
+    // Reduce bloating
+    if (indices_vec.empty() && set_data_it->second.test_elements_count == 0) {
+        node_data->set_data.erase(set_data_it);
+    }
+}
+
+void IsInNode::propagate(State& state) const {
+    const std::span<const Update> test_elements_diff = test_elements_ptr_->diff(state);
+    const std::span<const Update> element_diff = element_ptr_->diff(state);
+
+    if (test_elements_diff.empty() && element_diff.empty()) {
+        return;  // Nothing to do
+    }
+
+    IsInNodeData* node_data = data_ptr<IsInNodeData>(state);
+
+    // Save a copy of the updates for IsInNode::revert()
+    node_data->test_element_updates.assign(test_elements_diff.begin(), test_elements_diff.end());
+    node_data->element_updates.assign(element_diff.begin(), element_diff.end());
+
+    // Handle changes to `test_elements`
+    for (const Update& update : test_elements_diff) {
+        if (!update.placed()) {
+            // i.e. changed or removed.
+            ssize_t& count = node_data->set_data[update.old].test_elements_count;
+            // `update.old` must be in `test_elements_counter`
```

Comment:
I think this could go in the assert like `assert(count > 0 && "update.old should be in test_elements_counter")`